### PR TITLE
Fixes #254

### LINKED ISSUE #254: User Registration doesn't seem to be working well.
While trying to register a user from the browser on the url `/api/v1/users`, I get this error

```python
Internal Server Error: /api/v1/users/
Traceback (most recent call last):
  File "/home/curious/Dev/opensourceDev/care/care/users/models.py", line 139, in has_write_permission
    return request.data["user_type"] <= User.TYPE_VALUE_MAP["Volunteer"]
TypeError: '<=' not supported between instances of 'str' and 'int'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/viewsets.py", line 114, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 505, in dispatch
    response = self.handle_exception(exc)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 465, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 476, in raise_uncaught_exception
    raise exc
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 493, in dispatch
    self.initial(request, *args, **kwargs)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 411, in initial
    self.check_permissions(request)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 332, in check_permissions
    if not permission.has_permission(request, self):
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/dry_rest_permissions/generics.py", line 128, in has_permission
    return model_class.has_write_permission(request)
  File "/home/curious/Dev/opensourceDev/care/care/users/models.py", line 141, in has_write_permission
    return User.TYPE_VALUE_MAP[request.data["user_type"]] <= User.TYPE_VALUE_MAP["Volunteer"]
KeyError: '5'
ERROR 2020-04-04 16:36:23,767 log 4677 140162079278848 Internal Server Error: /api/v1/users/
Traceback (most recent call last):
  File "/home/curious/Dev/opensourceDev/care/care/users/models.py", line 139, in has_write_permission
    return request.data["user_type"] <= User.TYPE_VALUE_MAP["Volunteer"]
TypeError: '<=' not supported between instances of 'str' and 'int'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/viewsets.py", line 114, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 505, in dispatch
    response = self.handle_exception(exc)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 465, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 476, in raise_uncaught_exception
    raise exc
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 493, in dispatch
    self.initial(request, *args, **kwargs)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 411, in initial
    self.check_permissions(request)
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/rest_framework/views.py", line 332, in check_permissions
    if not permission.has_permission(request, self):
  File "/home/curious/.virtualenvs/corona-care/lib/python3.7/site-packages/dry_rest_permissions/generics.py", line 128, in has_permission
    return model_class.has_write_permission(request)
  File "/home/curious/Dev/opensourceDev/care/care/users/models.py", line 141, in has_write_permission
    return User.TYPE_VALUE_MAP[request.data["user_type"]] <= User.TYPE_VALUE_MAP["Volunteer"]
KeyError: '5'
```

### PR DESCRIPTION
**Pull request by abhiabhi94**
Change `user_type` received from the API before comparing

### CODE REVIEW

**Comment by vigneshhari** [comment]
The api cannot accept ENUM values now right ?

**Comment by abhiabhi94** [comment]
> e api cannot accept ENUM values now right ?
No, we were currently trying to search for a key `'5'` to the `TYPE_VALUE_CHOICES` dictionary while the accepted keys are something else.

**Comment by vigneshhari** [comment]
But in the old API , it could accept enum values right ?

**Comment by abhiabhi94** [comment]
I don't remember when enum was used for `user_type`. The latest development that I can remember was this type value mapping. Correct if I am wrong.

**Comment by vigneshhari** [comment]
in the earlier code , if there was a Type exception it was taking the value of the enum from the dictionary right ? but now it will create an error if it happens .

**Comment by abhiabhi94** [comment]
I have changed the type to integer before checking. So yes, I am expecting the API to return a string that can be matched. The TypeError was anyways, not handled correctly. We were searching for the passed value from the API among the keys in the type value mapping while the keys are different.

**Comment by abhiabhi94** [comment]
Should I change the choice mapping for all to enum here, if the changes are only required in models. I think it would be easy for everyone else to work. I am obviously assuming these choices are not used anywhere so the build is not broken.

**Comment by vigneshhari** [comment]
i can't confirm if the enums or integers are used anywhere in production, so let's be on the safe side and add options for registering with both ENUM and integers for now.

**Comment by anroopak** [comment]
@abhiabhi94 there is a Conflict, can you fix this?

**Comment by abhiabhi94** [comment]
@anroopak I have solved the conflict. What I don't understand if, as far I can see the openAPI documentation, `user_type` must be in integer form, but the data returned from `request.data['user_type']` was in string format.

File: care/users/models.py [hunk 1]
```diff
@@ -143,12 +143,16 @@ def has_object_read_permission(self, request):
...
+            user_type = int(request.data["user_type"])
+
+        except ValueError:
+            user_type = request.data["user_type"]
+
         except KeyError:
             # No user_type passed, the view shall raise a 400
             return True
+        finally:
+            return user_type <= User.TYPE_VALUE_MAP["Volunteer"]
```
Comment (vigneshhari) [inline care/users/models.py:155]: Codacy Issue found: [return statement in finally block may swallow exception](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5305091)

**Comment by abhiabhi94** [comment]
#277 fixes this too, so I'm closing this one.

**Pull request closed without merging**